accord-core/src/main/java/accord/local/durability/DurabilityQueue.java

```diff
@@ -73,325 +86,702 @@ public class DurabilityQueue
-        node.scheduler().selfRecurring(() -> submit(syncPoint, request, attempt), retryDelay, MICROSECONDS);
+        Invariants.require(done.deferred == null);
+        unregister(done);
+        done.status = DONE;
+        adapter.done(done.request, done.syncPoint);
     }
 
-    void restart(SyncPoint<Range> syncPoint, DurabilityRequest request, int attempt)
+    private void abandon(Pending abandon, boolean requeueDeferred, boolean pruned)
     {
-        long retryDelay = node.agent().retryDurabilityDelay(node, attempt, MICROSECONDS);
-        Invariants.require(retryDelay > 0);
-        logger.debug("{}: Restarting durability for {} in {}s.", syncPoint.syncId, syncPoint.route.toRanges(), String.format("%.3f", retryDelay/1000_000.0));
-        node.scheduler().selfRecurring(() -> node.durability().shards().request(request, request.stillWaiting(syncPoint.route)), retryDelay, MICROSECONDS);
+        Invariants.require(requeueDeferred || abandon.deferred == null);
+        Invariants.require(abandon.parent == null);
+        if (abandon.isInHeap())
+            queued.remove(abandon);
+        abandon.status = ABANDONED;
+        if (requeueDeferred)
+            requeueDeferred(abandon);
+        unregister(abandon);
+        adapter.abandon(abandon.request, abandon.syncPoint, pruned);
+    }
+
+    private synchronized void completing(Pending pending, long startedAt)
+    {
+        if (pending.status != ACTIVE || pending.startAt != startedAt)
+            return;
+
+        pending.status = COMPLETING;
+        --activeCount;
+        requeueDeferred(pending);
+        enqueue(requeue.poll());
+    }
+
+    private void requeueDeferred(Pending parent)
+    {
+        Invariants.require(parent.status.compareTo(COMPLETING) >= 0);
+        List<Pending> deferreds = parent.deferred;
+        if (deferreds == null)
+            return;
+
+        parent.deferred = null;
+        deferreds.sort(BY_PRIORITY);
+        int count = 0;
+        outer: for (int i = 0 ; i < deferreds.size() ; ++i)
+        {
+            Pending deferred = deferreds.get(i);
+            deferred.unsetParent();
+            for (int j = 0 ; j < count ; ++j)
+            {
+                Pending newParent = deferreds.get(j);
+                if (newParent.syncPoint.route.intersects((AbstractRanges) deferred.syncPoint.route))
+                {
+                    newParent.addDeferred(deferred);
+                    continue outer;
+                }
+            }
+            deferreds.set(count++, deferred);
+            requeue.add(deferred);
+        }
+    }
+
+    void requeue(Pending pending)
+    {
+        pending.unsetParent();
+        requeue.add(pending);
+    }
+
+    void setParent(List<Pending> deferred, Pending parent)
+    {
+        for (Pending p : deferred)
+        {
+            p.unsetParent();
+            p.parent = parent;
```

Comment:
Do we also need to add `deferred` to `deferred` on `parent`?

Reply:
I think this is no longer present